src/imperative.rkt

```diff
@@ -245,27 +260,44 @@
                  (equal? (language-name (*lang*)) "scala"))
              (or (not (equal? curr-prec new-prec))
                  (not (equal? curr-round new-round))))
-        (let-values ([(ctx* tmp-var) (ctx-random-name ctx)])
-          (unless (equal? curr-round new-round) 
-            (printf "~a~a\n" indent (change-round-mode new-round indent)))
-          (printf "~a~a\n" indent 
-            (convert-declaration (ctx-update-props ctx* props) tmp-var (convert-expr body #:ctx (ctx-update-props ctx* props) #:indent indent)))
-          (unless (equal? curr-round new-round)
-            (printf "~a~a\n" indent (change-round-mode curr-round indent)))
-          (if (equal? curr-prec new-prec)
-              tmp-var
-              (round-expr tmp-var ctx #t)))]
+        (define ctx* (ctx-update-props ctx props))
+        (define-values (body* body-prec) (convert-expr body #:ctx ctx* #:indent indent))
+        (if (and (equal? curr-round new-round) (equal? curr-prec body-prec))
+            (values body* body-prec)
+            (let-values ([(ctx** var) (ctx-random-name (ctx-update-props ctx* `(:precision ,body-prec)))])
+              (unless (equal? curr-round new-round) 
+                (printf "~a~a" indent (change-round-mode new-round indent)))
+              (printf "~a~a\n" indent (convert-declaration ctx** var body*))
+              (unless (equal? curr-round new-round)
+                (printf "~a~a" indent (change-round-mode curr-round indent)))
+              (if (equal? new-prec body-prec)
+                  (values var body-prec)
+                  (values (round-expr var ctx #t) body-prec))))]
       [else
        (convert-expr body #:ctx (ctx-update-props ctx props) #:indent indent)])]
 
     [(list (? operator? operator) args ...)
-      (define args_c
-          (map (λ (arg) (convert-expr arg #:ctx ctx #:indent indent)) args))
-     (convert-application ctx operator args_c)] 
-    [(list digits m e b) (convert-constant ctx (digits->number m e b))]
-    [(? constant?) (convert-constant ctx expr)]
-    [(or (? number?) (? hex?)) (convert-constant ctx expr)]
-    [(? symbol?) (ctx-lookup-name ctx expr)]))
+      (define-values (args* precs*)
+        (for/lists (args* precs*) ([arg args])
+          (convert-expr arg #:ctx ctx #:indent indent)))
+      (define cast? ; C implicit casting: check if arguments need to be explicitly casted
+        (and (equal? (language-name (*lang*)) "c")
+             (andmap (λ (x) (positive? (cmp-prec (ctx-lookup-prop ctx ':precision) x))) precs*)))
```

Comment:
Shouldn't it be the other way around, i.e., `negative?`?

This core
```
(FPCore exp32 (x)
  :precision binary64
  (! :precision binary32 (exp x)))
```
is translated to:
```c
double ex3(double x) {
	float tmp = expf(x);
	return tmp;
}
```

And this core
```
(FPCore exp64 (x)
  :precision binary32
  (! :precision binary64 (exp x)))
```
is translated to:
```c
float ex4(float x) {
	double tmp = exp((double) x);
	return tmp;
}
```
I think it makes more sense to convert `double` to `float` explicitly rather than `float` to `double`.

In general, I am not sure what is the right behavior in situations like this. The core `exp32` cannot be faithfully translated into C code. Maybe it is better to report errors for such cores.

Reply:
Oops, totally redundant. The reason for explicit casting was this example:
```
(FPCore add2 (x y)
(FPCore (x y)
  :precision binary64 
  (- (! :precision binary32 (+ x y)) 
     (! :precision binary32 (- x y))))
```
Without the cast the final type would be a `float`.